File: Facade/Facade.php
<?php

/**
 * Created by PhpStorm.
 * User: jamesyan
 * Date: 2017/3/23
 * Time: 22:08
 */
interface shape
{
    public function draw();
}

class Rectangle implements shape
{
    public function draw()
    {
        echo "Rectangle draw() method<br/>";
    }
}

class Square implements shape
{
    public function draw()
    {
        echo "Square draw() method<br/>";
    }
}

class Circle implements shape
{
    public function draw()
    {
        echo "Circle draw() method<br/>";
    }
}


class ShapeMaker
{
    private $circle, $rectangle, $square;

    public function __construct()
    {
        $this->circle = new Circle();
        $this->rectangle = new Rectangle();
        $this->square = new Square();
    }

    public function drawCircle(){
        $this->circle->draw();
    }

    public function drawRectangle(){
        $this->rectangle->draw();
    }

    public function drawSquare(){
        $this->square->draw();
    }
}

//demo
$shapeMaker=new ShapeMaker();
$shapeMaker->drawCircle();
$shapeMaker->drawRectangle();
$shapeMaker->drawSquare();

File: Facade/Facade.go
package main

import "fmt"

//shape接口
type Shape interface {
	draw()
}

//矩形
type Rectangle struct {

}

func (s *Rectangle)draw() {
	fmt.Print("Rectangle draw() method\r\n");
}


//方形
type Square struct {

}

func (s *Square)draw() {
	fmt.Print("Square draw() method\r\n");
}

//圆形
type Circle struct {

}

func (s *Circle)draw() {
	fmt.Print("Circle draw() method\r\n");
}

type ShapeMaker struct {
	circle Shape
	rectangle Shape
	square Shape
}

func(s *ShapeMaker)drawCircle(){
	s.circle.draw()
}

func(s *ShapeMaker)drawRectangle(){
	s.rectangle.draw()
}

func(s *ShapeMaker)drawSquare(){
	s.square.draw()
}


func main(){
	shapeMaker:=&ShapeMaker{new(Circle),new(Rectangle),new(Square)}
	shapeMaker.drawCircle()
	shapeMaker.drawRectangle()
	shapeMaker.drawSquare()
}




